server/services/personService.ts

```diff
@@ -10,15 +10,76 @@ import type {
 } from '../data'
 import { PersonUtils } from '../utils'
 import type { Person } from '@accredited-programmes/models'
+import type { OffenceWithDetail } from '@accredited-programmes/ui'
 import type { Caseload, OffenderSentenceAndOffences } from '@prison-api'
 
+interface OffenceHistory {
+  additionalOffences: Array<OffenceWithDetail>
+  indexOffence?: OffenceWithDetail
+}
+
 export default class PersonService {
   constructor(
     private readonly hmppsAuthClientBuilder: RestClientBuilderWithoutToken<HmppsAuthClient>,
     private readonly prisonApiClientBuilder: RestClientBuilder<PrisonApiClient>,
     private readonly prisonerSearchClientBuilder: RestClientBuilder<PrisonerSearchClient>,
   ) {}
 
+  async getOffenceHistory(token: Express.User['token'], prisonNumber: Person['prisonNumber']): Promise<OffenceHistory> {
+    try {
+      const prisonApiClient = this.prisonApiClientBuilder(token)
+
+      const offender = await prisonApiClient.findOffenderBookingByOffenderNo(prisonNumber)
+
+      if (!offender.offenceHistory) {
+        throw createError(404)
+      }
+
+      const offenceHistoryWithDetail: Array<OffenceWithDetail> = await Promise.all(
+        offender.offenceHistory.map(async offence => {
+          const offencesByCode = offence.offenceCode
+            ? await prisonApiClient.findOffencesThatStartWith(offence.offenceCode)
+            : undefined
+          const offenceDetail = offencesByCode?.content?.[0]
+
+          return {
+            code: offence.offenceCode,
+            date: offence.offenceDate,
+            description: offenceDetail?.description,
+            mostSerious: offence.mostSerious,
+            statuteCodeDescription: offenceDetail?.statuteCode?.description,
+          }
+        }),
+      )
+
+      return {
```

Comment:
I'm fine with the `filter.sort` and `.find` combination in the `return` for readability, but in case we want to keep the loops to a minimum, a `forEach` before the `return` could do the job (pushing to an array, and setting the `indexOffence` when `mostSerious` is found)

Reply:
I rarely (if ever) use a `forEach` as I _believe_ the individual array methods like filter and find are generally quicker. Tried a little test on jsperf for example https://jsperf.app/zujofe

Reply:
Whaaaaaat!? Mind blown